src/Validation/Validator.php

```diff
@@ -95,50 +137,224 @@ private function validateParams(array $params, string $formId): array
 	}
 
 	/**
-	 * Validate files.
+	 * Validate files from the validation reference.
 	 *
 	 * @param array<string, mixed> $files Files to check.
-	 * @param array<string, mixed> $params Params for reference.
+	 * @param array<int|string, mixed> $validationReference Validation reference to check against.
 	 * @param string $formId Form Id.
 	 *
 	 * @return array<int|string, string>
 	 */
-	private function validateFiles(array $files, array $params, string $formId = ''): array
+	private function validateFiles(array $files, array $validationReference, string $formId = ''): array
 	{
 		$output = [];
 
+		// Check files.
 		foreach ($files as $fileKey => $fileValue) {
-			$input = $params[$fileKey] ?? [];
+			// Find validation reference by ID.
+			$reference = $validationReference[$fileKey] ?? [];
 
-			if (!$input) {
+			// Bailout if no validation is required.
+			if (!$reference) {
 				continue;
 			}
 
-			$fileName = $fileValue['name'] ?? '';
-			$fileSize = $fileValue['size'] ?? '';
-
-			$inputDetails = json_decode($input, true);
-			$inputData = $inputDetails['data'] ?? [];
-			$inputName = $inputDetails['name'] ?? '';
-
-			foreach ($inputData as $dataKey => $dataValue) {
-				switch ($dataKey) {
-					case 'validationAccept':
-						if (!empty($dataValue) && !$this->isFileTypeValid($fileName, $dataValue) && $inputData['validationRequired'] === '1') {
-							$output[$inputName] = sprintf($this->labels->getLabel('validationAccept', $formId), $dataValue);
+			// Loop all validations from the reference.
+			foreach ($reference as $dataKey => $dataValue) {
+				// Check validation for accept file types.
+				if ($dataKey === 'accept') {
+					foreach ($fileValue['name'] as $file) {
+						if (!empty($dataValue) && !$this->isFileTypeValid($file, $dataValue)) {
+							$output[$fileKey] = sprintf($this->labels->getLabel('validationAccept', $formId), $dataValue);
+							continue;
 						}
-						break;
-					case 'validationMinSize':
-						if (!empty($dataValue) && !$this->isFileMinSizeValid((int) $fileSize, (int) $dataValue * 1000) && $inputData['validationRequired'] === '1') {
-							$output[$inputName] = sprintf($this->labels->getLabel('validationMinSize', $formId), $dataValue);
+					}
+				}
+
+				// Check validation for size min/max.
+				foreach ($fileValue['size'] as $file) {
+					// Check validation for min size.
+					if ($dataKey === 'minSize') {
+						if (!empty($dataValue) && !$this->isFileMinSizeValid((int) $file, (int) $dataValue * 1000)) {
```

Comment:
What can the `$file` be? Something other than integer? Because you're checking the size, but pass the file. Naming is confusing me.

Reply:
fixed naming